docs/structured-streaming-programming-guide.md

```diff
@@ -0,0 +1,1156 @@
+
+Imagine the quick example is modified and the stream contains lines along with the time when the line was generated. Instead of running word counts, we want to count words within 10 minute windows, updating every 5 minutes. That is, word counts in words received between 10 minute windows 12:00 - 12:10, 12:05 - 12:15, 12:10 - 12:20, etc. Note that 12:00 - 12:10 means data that arrived after 12:00 but before 12:10. Now, consider a word that was received at 12:07. This word should increment the counts corresponding to two windows 12:00 - 12:10 and 12:05 - 12:15. So the counts will be indexed by both, the grouping key (i.e. the word) and the window (can be calculated from the event-time).
+
+The result tables would look something like the following.
+
+![Window Operations](img/structured-streaming-window.png)
+
+Since this windowing is similar to grouping, in code, you can use `groupBy()` and `window()` operations to express windowed aggregations.
+
+<div class="codetabs">
+<div data-lang="scala"  markdown="1">
+
+{% highlight scala %}
+// Number of events in every 1 minute time windows
+df.groupBy(window(df.col("time"), "1 minute"))
+  .count()
+
+
+// Average number of events for each device type in every 1 minute time windows
+df.groupBy(
+     df.col("type"),
+     window(df.col("time"), "1 minute"))
+  .avg("signal")
+{% endhighlight %}
+
+</div>
+<div data-lang="java"  markdown="1">
+
+{% highlight java %}
+import static org.apache.spark.sql.functions.window;
+
+// Number of events in every 1 minute time windows
+df.groupBy(window(df.col("time"), "1 minute"))
+  .count();
+
+// Average number of events for each device type in every 1 minute time windows
+df.groupBy(
+     df.col("type"),
+     window(df.col("time"), "1 minute"))
+  .avg("signal");
+
+{% endhighlight %}
+
+</div>
+<div data-lang="python"  markdown="1">
+{% highlight python %}
+from pyspark.sql.functions import window
+
+# Number of events in every 1 minute time windows
+df.groupBy(window("time", "1 minute")).count()
+
+# Average number of events for each device type in every 1 minute time windows
+df.groupBy("type", window("time", "1 minute")).avg("signal")
+{% endhighlight %}
+
+</div>
+</div>
+
+
+Now consider what happens if one of the events arrives late to the application.
+For example, a word that was generated at 12:04 but it was received at 12:11. 
+Since this windowing is based on the time in the data, the time 12:04 should considered for windowing. This occurs naturally in our window-based grouping --the late data is automatically placed in the proper windows and the correct aggregates updated as illustrated below.
```

Comment:
Couple of minor corrections.
1) the time 12:04, should *be* considered for windowing. 
2) grouping the late data